Textarea sequence (#250)
ebi-webcomponents/nightingale: packages/nightingale-textarea-sequence/README.md

# nightingale-textarea-sequence

[![Published on NPM](https://img.shields.io/npm/v/@nightingale-elements/nightingale-textarea-sequence.svg)](https://www.npmjs.com/package/@nightingale-elements/nightingale-textarea-sequence)

A custom element that creates a formatted text area to capture sequences. It uses [QuillJS](https://quilljs.com/) to format the textarea.

## Usage

```html
<nightingale-textarea-sequence
  id="track"
  min-sequence-length="5"
  width="800"
  height="400"
></nightingale-textarea-sequence>
```

## API reference

### Attributes

#### `alphabet?: 'protein'|'dna'|string (default: 'protein')`

Either a `string` explicitly listing the valid characters in the sequence or one of the predefined alphabets:

- `"protein"`: `"ACDEFGHIKLMNPQRSTVWY "`
- `"dna"`: `"AGTCN "`

#### `case-sensitive?: boolean (default: false)`

Indicates if the checks against the alphabet should consider the sequence casing

#### `single?: boolean (default: false)`

Indicates if the textarea should only allow a single sequence

#### `disable-header-check?: boolean (default: false)`

Indicates if the checks against the alphabet should consider the absence of the header.

> This will only makes sense if the attribute `single` is also `true`, if it's not, the value of

Reviewer: ```suggestion
This will only make sense if the attribute `single` is also `true`, if it's not, the value of
```

Usage Example:

```javascript
element.quill.on("text-change", (e) => {
  console.log(element.sequence);
});
```

### Exposed functions

We have exposed some functions tht can be use without having to load the web component:

#### `formatSequence(sequence, options={block: 10, line: 50})`

Splits a string into lines of length `line` with block of `block` length separated by white spaces.

parameters:

- `sequence`: type: `string`
- `options`: type `{block: number, line: number}`

Returns:

- `string`

Usage example:

```javascript
import { formatSequence } from "textarea-sequence";
const seq = "XXXXXXXXXXXXXXX";
formatSequence(seq, 4, 8); // 'XXXX XXXX\nXXXX XXX'
formatSequence(seq, 2, 8); // 'XX XX XX XX\nXX XX XX X'
formatSequence(seq, 5, 10); // 'XXXXX XXXXX\nXXXXX'
```

#### `cleanUpText(text,alphabet = alphabets.protein, caseSensitive = false, removeComments = true, single = true, format = formatSequence)`

Takes a sequence and transform it, applying the following heuristics:

> - removes any character that's not in the alphabet.
> - Generates a header if the sequence doesn't have one
> - formats the sequence using the given function. default: blocks of 10 chars separated with a white space and lines of 50 bases.
> - If `case_sensitive` is `true` cases mismatches are removed.
> - If `removeComments` is `true`, removes any line that starts with `;`
> - If `single` is `true`, removes any sequence after the first one.

Reviewer: ```suggestion
- removes any character not in the alphabet
- generates a header if the sequence does not already have one
- formats the sequence using the specified function. Default: blocks of 10 characters separated with a white space and lines of 50 characters
- if `case_sensitive` is `true`, case mismatches are removed
- if `removeComments` is `true`, lines starting with `;` are removed
- if `single` is `true`, the first sequence is retained and the subsequent are removed
```

When all of those things are executed, the string is formatted with the function `format()`.

> parameters:

Reviewer: ```suggestion
Parameters:
```
